File: eve_trader_local/goonmetrics_client.py
"""Client for the Goonmetrics APIs (gnf.lt).

"Goonmetrics" is a naming legacy from the original third-party tool - the
endpoints this client actually calls are gnf.lt's rehosting of it. Kept under
the same name as in the parent eve-trader repo (see SYNC.md), where every
caller and every doc already says "the Goonmetrics client".

Two separate endpoints live here, answering two different questions - don't
treat a number from one as interchangeable with a same-sounding number from
the other:

* `current_prices` (appraise.gnf.lt) - best bid/best ask *right now* for
  every type in one market. Used as the failsafe behind
  `ESIClient.structure_order_stats_bulk_or_goonmetrics` when the real
  structure order book is unreachable.
* `price_history`/`price_history_chunked` (goonmetrics.apps.gnf.lt) - daily
  *region* history (avg/min/max price and units moved) for a batch of
  type_ids. This is what answers "is a newly discovered candidate
  historically worth importing at all", asked against
  `TradingConfig.reference_region_id` - a region-wide average smooths out
  the noise a live order-book snapshot is full of.

The history endpoint returns XML like:
    <evec_api><result><rowset name="history">
      <type id="...">
        <history date="..." avgPrice="..." maxPrice="..." minPrice="..."
                 movement="..." numOrders="..."/>
      </type>
    </rowset></result></evec_api>
"""
from __future__ import annotations

import logging
import threading
import time
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass
from typing import Iterable, Optional
from xml.etree import ElementTree as ET

import requests

from .config import TRADING_CONFIG, TradingConfig

log = logging.getLogger(__name__)

USER_AGENT = "eve-trader-local"

# Module-level (not per-instance, and not class-level) TTL cache, keyed by
# `market`. A GoonmetricsClient is constructed fresh at each call site, so a
# per-instance cache would never actually hit; module scope is what survives
# those instantiations. (esi_client.py's caches are class-level for the same
# "fresh client per call" reason - either scope works there; here there is no
# state on the class worth hanging it off, so plain module dicts are the
# simpler shape. sde.py caches nothing at all: its fetches are explicit,
# user-initiated refreshes, not something a request path re-enters.)
#
# The payload is the full market as one multi-megabyte JSON dump (~11MB for
# Jita, measured over 30s on the parent repo's connection) - re-downloading
# it twice in the same minute because the user clicked two actions in a row
# is the exact cost this exists to remove. 60s keeps it fresh enough for a
# real buy/sell decision.
#
# Locked per market, not with one shared lock: the lock is held for the whole
# (multi-second) fetch so two callers racing on a cold cache serialize rather
# than both paying the download - but a single shared lock would also make an
# in-flight Jita fetch block a concurrent home-market fetch, which shares no
# cache entry with it at all.
_PRICES_CACHE_TTL = 60  # seconds
_prices_cache: dict[str, list["CurrentPrice"]] = {}
_prices_cache_at: dict[str, float] = {}
_prices_locks: dict[str, threading.Lock] = {}
_prices_locks_guard = threading.Lock()  # guards creation of a new per-market lock only, never held during a fetch


def _lock_for_market(market: str) -> threading.Lock:
    with _prices_locks_guard:
        if market not in _prices_locks:
            _prices_locks[market] = threading.Lock()
        return _prices_locks[market]


def clear_prices_cache() -> None:
    """Forces the next current_prices call (for every market) to re-fetch.
    Module-level state doesn't reset itself between tests - anything
    exercising the real cache path should call this rather than depend on
    test ordering."""
    with _prices_locks_guard:
        _prices_cache.clear()
        _prices_cache_at.clear()


@dataclass(frozen=True)
class HistoryPoint:
    region_id: int
    type_id: int
    date: str
    min_price: float
    max_price: float
    avg_price: float
    movement: float
    num_orders: int


@dataclass(frozen=True)
class CurrentPrice:
    """Current best buy/sell for one type in a market, as served by
    appraise.gnf.lt."""
    type_id: int
    updated: str
    buy: float
    sell: float


class GoonmetricsClient:
    def __init__(self, cfg: TradingConfig = TRADING_CONFIG):
        self.cfg = cfg
        self.session = requests.Session()
        self.session.headers.update({"User-Agent": USER_AGENT})

    def current_prices(self, market: str) -> list[CurrentPrice]:
        """Current best buy (max)/sell (min) for every type in `market` (e.g.
        "jita", or a player structure's market slug), sorted by type_id.

        Retries with backoff before giving up: this is a third-party, no-SLA
        server serving a multi-megabyte response, and normal response-time
        variance alone (not an outage) was enough to kill every caller of a
        bare single-attempt fetch in the parent repo.

        Cached module-wide for _PRICES_CACHE_TTL seconds - see that
        constant's own comment.
        """
        with _lock_for_market(market):
            cached_at = _prices_cache_at.get(market, 0.0)
            if market in _prices_cache and (time.time() - cached_at) < _PRICES_CACHE_TTL:
                return list(_prices_cache[market])

            url = f"{self.cfg.goonmetrics_appraise_base}/market/{market}/prices.json"
            last_exc: Optional[requests.RequestException] = None
            for attempt in range(1, 4):
                try:
                    resp = self.session.get(url, timeout=60)
                    resp.raise_for_status()
                    break
                except requests.RequestException as e:
                    last_exc = e
                    if attempt < 3:
                        time.sleep(attempt * 2)
            else:
                raise last_exc
            prices = [
                CurrentPrice(
                    type_id=item["typeID"],
                    updated=item["prices"]["updated"],
                    buy=item["prices"]["buy"]["max"],
                    sell=item["prices"]["sell"]["min"],
                )
                for item in resp.json()
            ]
            prices.sort(key=lambda p: p.type_id)
            _prices_cache[market] = prices
            _prices_cache_at[market] = time.time()
            return list(prices)

    def price_history(self, region_id: int, type_ids: Iterable[int]) -> list[HistoryPoint]:
        """Daily history for a batch of type_ids in one region. Never raises
        on a Goonmetrics failure - silently falls back to the slower
        per-type_id ESI history endpoint instead, so callers don't need their
        own fallback handling."""
        type_ids = list(type_ids)
        ids = ",".join(str(t) for t in type_ids)
        url = f"{self.cfg.goonmetrics_history_base}?region_id={region_id}&type_id={ids}"
        try:
            resp = self.session.get(url, timeout=30)
            resp.raise_for_status()
        except requests.RequestException as e:
            # Goonmetrics is a third-party community API with no SLA - fall
            # back to ESI's own (slower, one call per type_id, but
            # always-available) daily history rather than letting a
            # Goonmetrics outage take down candidate discovery entirely.
            log.warning("Goonmetrics price history unavailable (%s) - falling back to ESI per-type history.", e)
            return self._esi_price_history_fallback(region_id, type_ids)
        return _parse_history_xml(resp.text, region_id)

    def _esi_price_history_fallback(self, region_id: int, type_ids: list[int]) -> list[HistoryPoint]:
        """Refetches via ESI's /markets/{region_id}/history/ (esi_client.
        region_market_history), one call per type_id since ESI has no batch
        endpoint for this.

        Maps ESI's schema onto HistoryPoint: `movement` is ESI's own `volume`
        field (units traded that day, *not* an ISK value - confirmed live in
        the parent repo against ESI directly), so it passes straight through.
        That multiplication by average_price used to happen here and silently
        mixed unit-count and ISK-value figures in the same
        HistoryPoint.movement field depending on which source served the
        request."""
        from .esi_client import ESIClient, ESIError  # local import: keeps callers that never hit this fallback free of an esi_client<->goonmetrics_client coupling

        esi = ESIClient(self.cfg)
        points: list[HistoryPoint] = []
        for type_id in type_ids:
            try:
                rows = esi.region_market_history(region_id, type_id)
            except ESIError as e:
                log.warning("ESI price history fallback also failed for type_id %d (%s) - skipping it.", type_id, e)
                continue
            for r in rows:
                points.append(HistoryPoint(
                    region_id=region_id, type_id=type_id, date=r["date"],
                    min_price=r["lowest"], max_price=r["highest"], avg_price=r["average"],
                    movement=r["volume"], num_orders=r["order_count"],
                ))
        return points

    def price_history_chunked(self, region_id: int, type_ids: list[int],
                              chunk_size: int | None = None, max_workers: int = 6) -> list[HistoryPoint]:
        """Chunks are independent requests (each already batches chunk_size
        type_ids into one call) - fetching them concurrently cuts wall-clock
        time roughly by max_workers, since this is network-latency-bound.

        Each chunk is isolated in its own try/except (on top of
        price_history's own Goonmetrics->ESI fallback for plain request
        failures) so a single chunk hitting something neither of those
        handles - a malformed response, a parsing bug - just loses that one
        chunk's points instead of aborting the whole call and everything
        after it (matters most for a full candidate search spanning many
        chunks over several minutes)."""
        chunk_size = chunk_size or self.cfg.chunk_size
        chunks = [type_ids[i:i + chunk_size] for i in range(0, len(type_ids), chunk_size)]
        if not chunks:
            return []

        def _fetch(chunk: list[int]) -> list[HistoryPoint]:
            try:
                return self.price_history(region_id, chunk)
            except Exception:  # noqa: BLE001 - one bad chunk must not lose the rest
                log.exception("Price history chunk failed for region %d (%d ids) - skipping it.",
                              region_id, len(chunk))
                return []

        out: list[HistoryPoint] = []
        with ThreadPoolExecutor(max_workers=max_workers) as pool:
            for points in pool.map(_fetch, chunks):
                out.extend(points)
        return out


def _parse_history_xml(xml_text: str, region_id: int) -> list[HistoryPoint]:
    root = ET.fromstring(xml_text)
    points: list[HistoryPoint] = []
    for type_el in root.iter("type"):
        type_id = int(type_el.attrib["id"])
        for hist_el in type_el.findall("history"):
            points.append(HistoryPoint(
                region_id=region_id,
                type_id=type_id,
                date=hist_el.attrib["date"],
                min_price=float(hist_el.attrib["minPrice"]),
                max_price=float(hist_el.attrib["maxPrice"]),
                avg_price=float(hist_el.attrib["avgPrice"]),
                movement=float(hist_el.attrib["movement"]),
                num_orders=int(hist_el.attrib["numOrders"]),
            ))
    return points

File: android/app/src/main/kotlin/com/pappmichel/evetraderlocal/data/history/GoonmetricsClient.kt
package com.pappmichel.evetraderlocal.data.history

import android.util.Xml
import java.io.StringReader
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit
import kotlinx.coroutines.withContext
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import org.xmlpull.v1.XmlPullParser

const val GOONMETRICS_HISTORY_BASE = "https://goonmetrics.apps.gnf.lt/api/price_history/"
private const val USER_AGENT = "eve-trader-local-android (contact: set EVE_CONTACT_EMAIL)"

// Same default as the desktop build's config.py `TradingConfig.chunk_size`
// (type_ids per Goonmetrics price-history request). Not a TradingConfig
// field on this platform yet: nothing else on Android has a reason to tune
// it, and TradingConfig here deliberately only carries the fields a ported
// screen actually reads (see its own docstring).
const val HISTORY_CHUNK_SIZE = 25

// Chunks are independent requests, so they're fetched concurrently -
// `price_history_chunked`'s own `max_workers=6` thread pool, expressed the
// way CandidateDiscovery.kt already expresses bounded fan-out on this
// platform (a Semaphore-bounded set of coroutines, never an unbounded
// `async` per chunk against a third-party no-SLA host).
private const val HISTORY_CONCURRENCY = 6

class GoonmetricsError(message: String) : RuntimeException(message)

/** One day of region price history for one type - the same fields as the
 * desktop build's goonmetrics_client.py `HistoryPoint` dataclass. */
data class HistoryPoint(
    val regionId: Int,
    val typeId: Int,
    val date: String,
    val minPrice: Double,
    val maxPrice: Double,
    val avgPrice: Double,
    val movement: Double,
    val numOrders: Int,
)

/** Client for gnf.lt's rehosting of the "Goonmetrics" community price-history
 * API - the price_history half of the desktop build's
 * goonmetrics_client.py, and only that half.
 *
 * Kept as its own client rather than folded into EsiClient.kt for the same
 * reason the desktop build keeps the two files apart: a different host, a
 * different (XML, not JSON) response format, and a third-party no-SLA
 * server whose failure modes have nothing to do with ESI's. The
 * `current_prices` half of the desktop client is deliberately *not* ported -
 * that one is the failsafe behind a structure order-book lookup, a
 * different feature nothing on Android reads yet.
 *
 * Response shape (parsed with Android's built-in XmlPullParser - no new
 * Gradle dependency needed for XML, and the whole document is a flat
 * attribute-only tree):
 *
 *     <evec_api><result><rowset name="history">
 *       <type id="...">
 *         <history date="..." avgPrice="..." maxPrice="..." minPrice="..."
 *                  movement="..." numOrders="..."/>
 *       </type>
 *     </rowset></result></evec_api>
 *
 * Deliberate omission vs. desktop: no ESI-history fallback. The desktop
 * `price_history` silently falls back to ESI's per-type
 * /markets/{region}/history/ when Goonmetrics is unreachable, because there
 * it sits inside a long unattended candidate search that shouldn't die on a
 * third-party outage. Here the only caller is a user-pressed "Refresh" on
 * one screen, where a failure is immediately visible and immediately
 * retryable - so a fetch failure just surfaces as a caught exception in the
 * screen's status line, the same way every other live-fetch screen on this
 * platform already handles a network failure. Worth adding later if this
 * client picks up an unattended caller.
 *
 * Retry/backoff mirrors EsiClient.kt's `getBody` (which in turn mirrors
 * esi_client.py's `_get_response`): 429 and 5xx back off and retry,
 * anything else raises immediately. */
class GoonmetricsClient(
    private val http: OkHttpClient = OkHttpClient(),
    private val historyBase: String = GOONMETRICS_HISTORY_BASE,
) {
    /** Daily history for a batch of type_ids in one region, in one request -
     * the endpoint takes a comma-separated type_id list, which is why
     * chunking (below) is about request size, not about one call per id. */
    suspend fun priceHistory(regionId: Int, typeIds: List<Int>): List<HistoryPoint> {
        if (typeIds.isEmpty()) return emptyList()
        val body = getBody(
            mapOf("region_id" to regionId.toString(), "type_id" to typeIds.joinToString(",")),
        )
        return parseHistoryXml(body, regionId)
    }

    /** Splits `typeIds` into `chunkSize` batches and fetches them
     * concurrently - the direct counterpart of `price_history_chunked`.
     *
     * Each chunk is isolated in its own try/catch, same as the desktop
     * version's per-chunk try/except: one chunk failing (an outage that
     * outlasts the retries, a malformed response) loses that chunk's points
     * instead of the whole refresh. A caller that gets *nothing* back sees
     * that as an empty result, not as an exception - the screen's
     * "no item has enough paired history yet" message covers that case
     * honestly either way. */
    suspend fun priceHistoryChunked(
        regionId: Int,
        typeIds: List<Int>,
        chunkSize: Int = HISTORY_CHUNK_SIZE,
    ): List<HistoryPoint> = coroutineScope {
        val semaphore = Semaphore(HISTORY_CONCURRENCY)
        typeIds.chunked(chunkSize).map { chunk ->
            async(Dispatchers.IO) {
                semaphore.withPermit {
                    try {
                        priceHistory(regionId, chunk)
                    } catch (e: Exception) {
                        emptyList()
                    }
                }
            }
        }.flatMap { it.await() }
    }

    private suspend fun getBody(params: Map<String, String>, retries: Int = 3): String =
        withContext(Dispatchers.IO) {
            val urlBuilder = historyBase.toHttpUrl().newBuilder()
            params.forEach { (key, value) -> urlBuilder.addQueryParameter(key, value) }
            val request = Request.Builder().url(urlBuilder.build()).header("User-Agent", USER_AGENT).build()

            var lastError: String? = null
            for (attempt in 1..retries) {
                val response = http.newCall(request).execute()
                response.use {
                    if (it.isSuccessful) {
                        return@withContext it.body!!.string()
                    }
                    lastError = "HTTP ${it.code} for ${request.url}"
                    if (it.code == 429 || it.code in 500..504) {
                        if (attempt < retries) {
                            delay((attempt * 1500).toLong())
                            return@use
                        }
                    }
                    throw GoonmetricsError(lastError!!)
                }
            }
            throw GoonmetricsError(lastError ?: "Exhausted retries for price history")
        }
}

/** Pull-parses the history document into flat HistoryPoints, carrying
 * `regionId` in from the request (the response body doesn't repeat it) -
 * same as `_parse_history_xml`'s own `region_id` parameter.
 *
 * Any `<history>` row missing a numeric attribute is skipped rather than
 * failing the whole document: one malformed day shouldn't cost the caller
 * every other type_id in the same chunk. Top-level, not a method, so it can
 * be tested against a hand-written document with no client or network
 * involved (see GoonmetricsClientTest) - which is also why the parser
 * itself is a defaulted parameter: `android.util.Xml` is a framework stub
 * on a plain JVM unit-test classpath and throws if called, so the test
 * passes in a real XmlPullParser implementation instead of mocking the
 * framework. */
fun parseHistoryXml(xml: String, regionId: Int, parser: XmlPullParser = Xml.newPullParser()): List<HistoryPoint> {
    parser.setInput(StringReader(xml))

    val points = mutableListOf<HistoryPoint>()
    var typeId: Int? = null
    var event = parser.eventType
    while (event != XmlPullParser.END_DOCUMENT) {
        when (event) {
            XmlPullParser.START_TAG -> when (parser.name) {
                "type" -> typeId = parser.getAttributeValue(null, "id")?.toIntOrNull()
                "history" -> {
                    val point = historyPoint(parser, regionId, typeId)
                    if (point != null) points.add(point)
                }
            }
            XmlPullParser.END_TAG -> if (parser.name == "type") typeId = null
        }
        event = parser.next()
    }
    return points
}

private fun historyPoint(parser: XmlPullParser, regionId: Int, typeId: Int?): HistoryPoint? {
    if (typeId == null) return null
    val date = parser.getAttributeValue(null, "date") ?: return null
    val minPrice = parser.getAttributeValue(null, "minPrice")?.toDoubleOrNull() ?: return null
    val maxPrice = parser.getAttributeValue(null, "maxPrice")?.toDoubleOrNull() ?: return null
    val avgPrice = parser.getAttributeValue(null, "avgPrice")?.toDoubleOrNull() ?: return null
    val movement = parser.getAttributeValue(null, "movement")?.toDoubleOrNull() ?: return null
    val numOrders = parser.getAttributeValue(null, "numOrders")?.toIntOrNull() ?: return null
    return HistoryPoint(
        regionId = regionId, typeId = typeId, date = date, minPrice = minPrice,
        maxPrice = maxPrice, avgPrice = avgPrice, movement = movement, numOrders = numOrders,
    )
}
